library/alloc/src/borrow.rs

```diff
@@ -324,6 +324,63 @@ impl<B: ?Sized + ToOwned> Cow<'_, B> {
     }
 }
 
+#[unstable(feature = "cow_disown", issue = "none")]
+impl<'a, B: ?Sized + ToOwned> Cow<'a, B> {
+    /// Replaces an owned value with the given borrowed value if it satisfies a
+    /// condition. Returns the owned value if any.
+    ///
+    /// If `func` returns `true`, the owned value is replaced with `sub` and
+    /// returned. Otherwise, `None` is returned.
+    ///
+    /// # Example
+    /// Replace an owned string if its lowercase form is equal to `"moo"`.
+    /// ```
+    /// let mut cow = Cow::Owned(String::from("Moo"));
+    /// let moo = "moo";
+    /// assert_eq!(cow.replace_if(moo, |o| o.to_lowercase() == moo), Some(String::from("Moo")));
+    /// assert_eq!(cow.replace_if(moo, |o| o.to_lowercase() == moo), None);
+    /// assert_eq!(cow, Cow::Borrowed("moo"));
+    /// ```
+    #[unstable(feature = "cow_disown", issue = "none")]
+    pub fn disown_if<'b: 'a>(
+        &mut self,
+        sub: &'b B,
+        func: impl FnOnce(&<B as ToOwned>::Owned) -> bool,
```

Comment:
Can you elaborate on why this takes a `&Owned` instead of a `&B`?

It's somewhat weird to me that this only compares in the owned case.  Replacing an owned "mOO" with borrowed "Moo" but leaving a borrowed "mOO" alone feels really fragile.

Reply:
Because the owned value is what's being checked. I just figured being as specific and unassuming as possible would be the safest bet.

> It's somewhat weird to me that this only compares in the owned case.

Well if nothing is owned, then there's nothing to be disowned. The behavior you describe seems better fit for something like `cow.replace_if()`.